Source/Core/VideoCommon/VertexLoaderManager.cpp

```diff
@@ -72,16 +72,32 @@ void UpdateVertexArrayPointers()
   if (!g_main_cp_state.bases_dirty)
     return;
 
+  // Only update the array base if the vertex description states we are going to use it.
+  auto is_indexed = [](VertexComponentFormat f) {
+    return f == VertexComponentFormat::Index8 || f == VertexComponentFormat::Index16;
+  };
+
   // Some games such as Burnout 2 can put invalid addresses into
   // the array base registers. (see issue 8591)
   // But the vertex arrays with invalid addresses aren't actually enabled.
   // Note: Only array bases 0 through 11 are used by the Vertex loaders.
   //       12 through 15 are used for loading data into xfmem.
-  for (int i = 0; i < 12; i++)
+  if (is_indexed(g_main_cp_state.vtx_desc.low.Position))
+    cached_arraybases[0] = Memory::GetPointer(g_main_cp_state.array_bases[0]);
+
+  if (is_indexed(g_main_cp_state.vtx_desc.low.Normal))
+    cached_arraybases[1] = Memory::GetPointer(g_main_cp_state.array_bases[1]);
+
+  for (size_t i = 0; i < g_main_cp_state.vtx_desc.low.Color.Size(); i++)
+  {
+    if (is_indexed(g_main_cp_state.vtx_desc.low.Color[i]))
+      cached_arraybases[i + 2] = Memory::GetPointer(g_main_cp_state.array_bases[i + 2]);
+  }
+
+  for (size_t i = 0; i < g_main_cp_state.vtx_desc.high.TexCoord.Size(); i++)
   {
-    // Only update the array base if the vertex description states we are going to use it.
-    if (g_main_cp_state.vtx_desc.GetVertexArrayStatus(i) & MASK_INDEXED)
-      cached_arraybases[i] = Memory::GetPointer(g_main_cp_state.array_bases[i]);
+    if (is_indexed(g_main_cp_state.vtx_desc.high.TexCoord[i]))
+      cached_arraybases[i + 4] = Memory::GetPointer(g_main_cp_state.array_bases[i + 4]);
```

Comment:
The different array numbers (0, 1, 2, 4, also 12 (not seen here)) should probably be extracted into constants.

Reply:
Looks like they actually are already constants, just not ones I used:

https://github.com/dolphin-emu/dolphin/blob/79ee6aae0a6a7add698bb7fbc0ef25aa166ccdec/Source/Core/VideoCommon/CPMemory.h#L10-L18